lib/puppetserver/ca/action/list.rb

```diff
@@ -60,16 +71,24 @@ def run(input)
           puppet = Config::Puppet.parse(config)
           return 1 if Errors.handle_with_usage(@logger, puppet.errors)
 
-          all_certs = get_all_certs(puppet.settings)
-          return 1 if all_certs.nil?
+          filter_names = certnames.any? \
+            ? lambda { |x| certnames.include?(x['name']) }
+            : lambda { |x| true }
 
+          all_certs = get_all_certs(puppet.settings).select { |cert| filter_names.call(cert) }
           requested, signed, revoked = separate_certs(all_certs)
-          input['all'] ? output_certs_by_state(requested, signed, revoked) : output_certs_by_state(requested)
+          missing = certnames - all_certs.map { |cert| cert['name'] }
+
+          (all || certnames.any?) \
+            ? output_certs_by_state(requested, signed, revoked, missing)
+            : output_certs_by_state(requested)
 
-          return 0
+          return missing.any? \
+            ? 1
+            : 0
         end
 
-        def output_certs_by_state(requested, signed = [], revoked = [])
+        def output_certs_by_state(requested, signed = [], revoked = [], missing = [])
           if revoked.empty? && signed.empty? && requested.empty?
```

Comment:
If you only input one certname and it's missing, this will be true and we will return, not printing the warning about missing certs.